include/RAJA/exec-cuda/reduce_cuda.hxx

```diff
@@ -931,104 +955,111 @@ class ReduceMinLoc<cuda_reduce<BLOCK_SIZE>, T> {
     __shared__ CudaReductionLocBlockDataType sd[BLOCK_SIZE];
     __shared__ bool lastBlock;
 
+    int blockId = blockIdx.x 
+       + blockIdx.y * gridDim.x 
+       + gridDim.x * gridDim.y * blockIdx.z; 
+    int threadId = threadIdx.x + 
+               blockDim.x * threadIdx.y + 
+               (blockDim.x * blockDim.y) * threadIdx.z;
     // initialize shared memory
     for (int i = BLOCK_SIZE / 2; i > 0; i /= 2) {
       // this descends all the way to 1
-      if (threadIdx.x < i) {
+      if (threadId < i) {
         // no need for __syncthreads()
-        sd[threadIdx.x + i].val = m_reduced_val;
-        sd[threadIdx.x + i].idx = m_reduced_idx;
+        sd[threadId + i].val = m_reduced_val;
+        sd[threadId + i].idx = m_reduced_idx;
       }
     }
     __syncthreads();
 
-    sd[threadIdx.x].val = val;
-    sd[threadIdx.x].idx = idx;  // need to reconcile loc vs idx naming
+    sd[threadId].val = val;
+    sd[threadId].idx = idx; 
     __syncthreads();
 
     for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
-      if (threadIdx.x < i) {
-        sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + i]);
+      if (threadId < i) {
+        sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + i]);
       }
       __syncthreads();
     }
 
-    if (threadIdx.x < 16) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 16]);
+    if (threadId < 16) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 16]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 8) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 8]);
+    if (threadId < 8) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 8]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 4) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 4]);
+    if (threadId < 4) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 4]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 2) {
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 2]);
+    if (threadId < 2) {
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 2]);
     }
     __syncthreads();
 
-    if (threadIdx.x < 1) {
+    if (threadId < 1) {
       lastBlock = false;
-      sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x], sd[threadIdx.x + 1]);
-      m_blockdata[m_blockoffset + blockIdx.x + 1] =
-          RAJA_MINLOC(sd[threadIdx.x],
-                      m_blockdata[m_blockoffset + blockIdx.x + 1]);
+      sd[threadId] = RAJA_MINLOC(sd[threadId], sd[threadId + 1]);
+      m_blockdata[m_blockoffset + blockId + 1] =
+          RAJA_MINLOC(sd[threadId],
+                      m_blockdata[m_blockoffset + blockId + 1]);
+      __threadfence();
```

Comment:
Its threadId == 0 calling threadfence. So it's only one thread here

I can put volatile on m_blockdata -- which as you imply causes the compiler to use a store that bypasses L1.  Just for discussion sake It seems like if this were the case we would not need a threadfence to begin with. As a fence ensure that the global update is finished before I call the atomic increment and fall thru and then the "last" block has assurances the right values are stored.

Without it I've seen some rare races which will cause the unit test to fail otherwise. So the fence seems good here ;)